keps/sig-cli/kustomize-exec-secret-generator.md

```diff
@@ -0,0 +1,289 @@
+
+```yaml
+secretGenerator:
+- name: fromPlugins
+  kvSources:
+  - pluginType: exec
+    name: my-exec-plugin
+    args:
+    - anotherArg
+    - yetAnotherArg
+```
+
+```sh
+$ kustomize build ./
+```
+
+Output:
+
+```bash
+secretGenerator plugins used, but plugins not enabled.  To enable plugins for trusted sources, specify --enable-plugins.
+secretGenerator plugins that will be run with --enable-plugins:
+$XDG_CONFIG_HOME/kustomize/plugins/builtin/my-exec-plugin anotherAg yetAnotherARg
+```
+
+## Risks and Mitigations
+
+### Security
+
+**Risk:** Chuck is able to exploit this feature to do something bad on Alice's machine
+
+Required steps to exploit:
+
+- Alice executes (via `-k` or `kustomize`) a malicious `kustomization.yaml`
+  - Alice does *not* run *without* enabling plugins with `--enable-plugins` to see which
+    plugins will be run and with which values.
+- Alice opted-in to enable plugins by providing the flag `--enable-plugins`
+- Alice or her organization installed the targeted SecretGenerator to `$XDG_CONFIG_HOME/kustomize/plugins/kvSources`
+  or another location she provided with `--plugin-path=<path-including-exploitable-binary>`
+- The specified SecretGenerator must be exploitable
+  - Simple transformation functions would not be exploitable.
+  - e.g. A command like `cat` would not be possible for Chuck to exploit in a meaningful way.
+
+Example Exploit Commands:
+
+In kustomize:
+
+```sh
+# User enables the plugin and allows it to exec sh
+$ kustomize build https://github.com/chuck/maliciousapp --enable-plugins --plugin-path=/bin
+```
+
+In kubectl:
+
+```sh
+# User enables the plugin and allows it to exec sh
+$ kubectl apply -k https://github.com/chuck/maliciousapp --enable-plugins --plugin-path=/bin
+```
+
+Analysis:
+
+This is a low risk profile.  Alice has to: kustomize a malicious kustomization file,
+explicitly enable plugins as command line arguments, and install an exploitable plugin to
+exploit the plugin system.
+
+By default kustomize will not-execute the plugins, and print out which plugins were specified
+with their arguments.
+
+## Alternatives Considered
+
+### Git Style Plugins
+
+Create a plugin mechanism similar to git or kubectl plugins which supports sub commands.
+
+kubectl and git plugins are targeted at providing a clean interactive UX with extensible sub commands.
+Because the plugins will be invoked declaratively rather than imperatively support for this sort of
+UX is not necessary.
+
+
+## Graduation Criteria
+
+Target Launch Status: Beta
+
+This is an integration of the previous execution SecretGenerator mechanism, and as
+such is relatively well understood.
+
+### Graduation to GA
+
+Gather user feedback.  Determine if there are addressable gaps.
+
+Consider the following:
+
+#### Enable Plugins with Environment Variables
+
+Allow users to specify `export KUSTOMIZE_ENABLE_PLUGINS=true` instead of providing the flag.
```

Comment:
Although it's convenient, env vars usually hide stuff. For this particular usecase I'd prefer we **do not** have env var. Although the alias problem (see my other comment) is still a thing.

Reply:
The env var acknowledges that the flag is merely a source of annoyance.